src/openMVG/sfm/pipelines/sequential/sequential_SfM.cpp

```diff
@@ -1416,7 +1414,12 @@ bool SequentialSfMReconstructionEngine::Resection(const std::size_t viewIndex)
         const Vec2 xJ = _features_provider->feats_per_view.at(J)[track.at(J)].coords().cast<double>();
 
         // test if the track already exists in 3D
-        if (_sfm_data.structure.count(trackId) != 0)
+        bool trackIdExists;
+#pragma omp critical
```

Comment:
this solves the hang problem but I'm afraid it leaves open a problem of data integrity.
Suppose:
* thread 1 enter the critical section, its point does not exists, so trackIdExists is false, it exits the critical region and now it is ready to go in the `else` branch to triangulate and create the point.
* thread 2 take the lock for the critical section, its point is the same as thread 1, it checks but it does not exists, so again it is ready to hit the `else` branch but thats wrong! because the points will be create by the other one (or whichever get the critical first). 

So the risk is that one thread enters the else branch but the point has been created meanwhile by another thread. Which, for how the code is written here is not even much of a problem because of the 
```c++
// Add a new track
Landmark & landmark = _sfm_data.structure[trackId];
landmark.X = X_euclidean;
landmark.obs[I] = Observation(xI, track.at(I));
landmark.obs[J] = Observation(xJ, track.at(J));
```
which basically takes an existing point (believing its a new point) update the point coordinates and the observation list. Very dangerous anyway!

As I told you I wrote a simplified version of this code that preserves the code architecture 
https://gist.github.com/simogasp/12fbd37d5241d9fb3df8aab0efc3c650
It takes as input a sfmdata file, and it basically tries to build a copy of the Landmarks in it by running through all the views and their observations, and adding either a new point to the copy if it does not exist or just updating the observations. It mimics this code minus the triangulation and other stuff.
If you run it with openMP you'll see in the end that in general some 3D points have less observations than the orginal. And above all, you'll see that `new_added_track` is bigger than the size of Landmarks, meaning that there are points that have been attempted to be added more than 1 time.

The safer solution seems to be putting the `critical` for the whole section, this guarantees the integrity and the hang problem (but of course performances are reduced, and it will be maybe slower than the single-threaded version...)